### DISCUSSION TITLE
Discriminator is unable to adapt to high loss.

### DISCUSSION THREAD

**Discussion by VirtualNonsense** [tracel-ai/burn#2921] [Q&A] [answered]
I'm currently trying to implement a pix2pix cgan to translate images into sketches ([repository](https://github.com/VirtualNonsense/sketchy_pix2pix)).
I tried setting up the network according to this [keras tutorial](https://machinelearningmastery.com/how-to-develop-a-pix2pix-gan-for-image-to-image-translation/) however I'm running a bit against a brick wall since I'm not able to get the training to work properly.

My issue seems to be that the discriminator is not able to adjust when i [punish it with the success of the generator](https://github.com/VirtualNonsense/sketchy_pix2pix/blob/92d1db86cd6419126b8f11c1bb849806b9e7870b/crates/sketchy_pix2pix/src/pix2pix/gan.rs#L137). The loss just keeps ramping up until the discriminator classifies everything as valid picture.
![image](https://github.com/user-attachments/assets/7a845719-f64a-4aab-822a-5fce79c84277)

![image](https://github.com/user-attachments/assets/215081ed-e529-4d70-8b74-30e642c58120)

Is there a way to retrieve tensors from gradients? 
I suspect this might be due to a local minimum that i have to find my way out of, but I would like to have a clearer insight into whats going on.

**Comment by laggui** [comment]
Sorry for the delayed response!

> Is there a way to retrieve tensors from gradients?

With the `GradientsParams`, you can retrieve the gradients for a given `ParamId` (associated to a parameter tensor) with `grads.get(id)`.

**Reply by laggui** [reply]
Just saw your issue in #2924, so looks like you at least figured this part out 😅 

How are you using the `GradientParams`? You could implement a module mapper similar to how the optimizers are [implemented](https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/optim/simple/adaptor.rs#L99).

In `map_float` you could inspect the gradients.

**Reply by VirtualNonsense** [reply]
Thank you for your answer!

I just answered to your response within that issue!

> How are you using the GradientParams? You could implement a module mapper similar to how the optimizers are [implemented](https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/optim/simple/adaptor.rs#L99).
In map_float you could inspect the gradients.

I will have a look!

**Comment by VirtualNonsense** [comment, marked as answer]
I think i might have resolved at least parts of my issue:
When training two model graphs in parallel one has to be quite particular when to detach a tensor from the model graph.

In my naive approach i detached the tensor while calculating the loss. If I understand it correctly this cause the tensor to be disconnected from the operations before and therefore the optimizer is not able to "reach" this operations in the next optimization step to change the corresponding weights.
Here is what [i did](https://github.com/VirtualNonsense/sketchy_pix2pix/commit/e447a4c8959b5de988ab81e0f359d26cc9b5fc42#diff-84cf2020d7679215b8439d32fc828c5eb84eefaaf58c985ca17dd52fc94f6547L114) it to fix the issue.

**Reply by laggui** [reply]
Ahh ok, glad you figured it out.

That's correct, `tensor.detach()` will detach the tensor from the autodiff graph so previous operations are not considered for autodiff.

Gotta be careful about that 😄
